src/library/scala/collection/mutable/TreeSet.scala

```diff
@@ -29,88 +28,145 @@ object TreeSet extends MutableSortedSetFactory[TreeSet] {
+    private[this] def pickLowerBound(newFrom: Option[A]): Option[A] = (from, newFrom) match {
+      case (Some(fr), Some(newFr)) => Some(ordering.max(fr, newFr))
+      case (None, _) => newFrom
+      case _ => from
+    }
 
-  override def contains(elem: A): Boolean = {
-    def leftAcceptable: Boolean = from match {
-      case Some(lb) => ordering.gteq(elem, lb)
-      case _ => true
+    /**
+     * Given a possible new upper bound, chooses and returns the most constraining one (the minimum).
+     */
+    private[this] def pickUpperBound(newUntil: Option[A]): Option[A] = (until, newUntil) match {
+      case (Some(unt), Some(newUnt)) => Some(ordering.min(unt, newUnt))
+      case (None, _) => newUntil
+      case _ => until
     }
 
-    def rightAcceptable: Boolean = until match {
-      case Some(ub) => ordering.lt(elem, ub)
-      case _ => true
+    /**
+     * Returns true if the argument is inside the view bounds (between `from` and `until`).
+     */
+    private[this] def isInsideViewBounds(key: A): Boolean = {
+      val afterFrom = from.isEmpty || ordering.compare(from.get, key) <= 0
+      val beforeUntil = until.isEmpty || ordering.compare(key, until.get) < 0
+      afterFrom && beforeUntil
     }
 
-    (notProjection || (leftAcceptable && rightAcceptable)) &&
-      RB.contains(treeRef.elem, elem)
-  }
+    override def rangeImpl(from: Option[A], until: Option[A]): TreeSet[A] =
+      new TreeSetView(pickLowerBound(from), pickUpperBound(until))
 
-  override def iterator: Iterator[A] = iteratorFrom(None)
+    override def contains(key: A) = isInsideViewBounds(key) && RB.contains(tree, key)
 
-  override def keysIteratorFrom(start: A) = iteratorFrom(Some(start))
+    override def iterator = RB.keysIterator(tree, from, until)
+    override def keysIteratorFrom(start: A) = RB.keysIterator(tree, pickLowerBound(Some(start)), until)
+    override def iteratorFrom(start: A) = RB.keysIterator(tree, pickLowerBound(Some(start)), until)
 
-  private def iteratorFrom(start: Option[A]) = {
-    val it = RB.keysIterator(treeRef.elem, pickBound(ordering.max, from, start))
-    until match {
-      case None => it
-      case Some(ub) => it takeWhile (k => ordering.lt(k, ub))
+    override def size = iterator.length
+    override def isEmpty = !iterator.hasNext
+
+    override def head = headOption.get
+    override def headOption = {
+      val elem = if (from.isDefined) RB.minKeyAfter(tree, from.get) else RB.minKey(tree)
+      (elem, until) match {
+        case (Some(e), Some(unt)) if ordering.compare(e, unt) >= 0 => None
+        case _ => elem
+      }
+    }
+
+    override def last = lastOption.get
+    override def lastOption = {
+      val elem = if (until.isDefined) RB.maxKeyBefore(tree, until.get) else RB.maxKey(tree)
+      (elem, from) match {
+        case (Some(e), Some(fr)) if ordering.compare(e, fr) < 0 => None
+        case _ => elem
+      }
     }
+
+    override def foreach[U](f: A => U): Unit = iterator.foreach(f)
```

Comment:
It does not scan the entire tree, only the relevant part:

https://github.com/scala/scala/blob/2.12.x/src/library/scala/collection/mutable/RedBlackTree.scala#L449

Double-check if changing to foreach is worth it with a benchmark, as Rex
suggests -- it might be that the speed difference is minor, since the
iterator does not use a stack, but the parent links.


On 7/10/2015 12:49 AM, Ichoran wrote:
>>      }
>> +
>> +    override def foreach[U](f: A => U): Unit = iterator.foreach(f)
> 
> Er, but I haven't looked at the implementation, so if that iterator needs to walk the whole tree in order to hit just the slice it cares about, I agree that an improvement is in order.  O(n) where n is big is really not fun when you think you have just a tiny (view of a) tree.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/scala/scala/pull/4608/files#r34312923
> 

-- 
Aleksandar Prokopec